Issues-1077 / Issue-1083 - Prefetch reads in-order / background flush writes in-order using `p-map-iterable`

huntharo opened the pull request:
## Release Suggestions
- [x] Merge PRs for Promises leak fix
- [x] Merge tests and fixes for stopping on errors #1089
- [x] Release contents of `master` at this point as this should have no dependency or speed changes
- [x] Wait for other PRs to merge:
   - [x] #1092 
   - [x] #1094 
   - [x] #1095 
   - [x] #1090 
- [x] Rebase this PR and resolve merge conflicts
- [x] Confirm tests are passing 
- [ ] Bump major version before releasing this PR - This PR will remove the 5 batches per 5 seconds default throttle and use only backpressure and `throttleInterval` to slow down the overall rate
- [ ] Merge and release this PR

## Changes
- Performance Summary
   - 3.5x to 6.4x faster with default settings for both, node 20, 7-27 ms RTT from ES
   - 2.5x to 3x faster with optimal settings for both, node 20, 7-27 ms RTT from ES
   - 10x to 20x faster with concurrency 10 vs defaults for baseline, node 20, 7-27 ms RTT from ES
- Compatible with Node v10
- Usage with Node v20 is highly recommended as it allows 2-3x higher performance vs v18
- Keep reads and writes in-order by default
- Does not allow changing reads to out-of-order (as ES scrolls and files are both incompatible with that)
- `concurrency` option (if value > 1) allows out of order writes with up to `concurrency` in-progress writes before applying backpressure - can be used with ES destination but should not be used with files
- Removes 5 batches per 5 seconds behavior that was being imposed by default settings of `queue`
   - `carryoverConcurrencyCount` - No longer does anything
   - `intervalCap` - No longer does anything
   - `concurrencyInterval` - No longer does anything
- Throttling is entirely controlled by:
   - `throttleInterval`
   - Speed of reads and RTT
   - Speed of writes and RTT
   - Backpressure of writes (typically the longest operation)
- Fixes #1077 
- Fixes #1083

## Performance

### Table of Results

Note: 6.117.0 was not re-run with `node 20` because the limiting factor was the default `--intervalCap=5` which held the throughput rate to 1 batch every second.

| Configuration | Node Version | Records Written | Duration (s) | Avg RTT (ms) | Records / Batch | Batches / Second |
|--------------|--------------|----------------:|-------------:|-------------:|--------------:|---------------:|
| PR (concurrency=10, limit=1000, esCompress=true) | 20.13.1 | 66,000 | 30 | 7 | 1,000 | 2.2 |
| PR (concurrency=10) | 20.13.1 | 61,400 | 30 | 7 | 100 | 20.5 |
| PR (concurrency=10, limit=1000, esCompress=true) | 20.13.1 | 56,000 | 30 | 26.7 | 1,000 | 1.9 |
| PR (concurrency=10) | 20.13.1 | 30,500 | 30 | 26.7 | 100 | 10.2 |
| PR (concurrency=10) | 18.16.1 | 13,000 | 30 | 26.7 | 100 | 4.3 |
| PR (concurrency=1) | 20.13.1 | 19,200 | 30 | 7 | 100 | 6.4 |
| PR (concurrency=1) | 20.13.1 | 10,500 | 30 | 26.7 | 100 | 3.5 |
| PR (concurrency=1) | 18.16.1 | 6,300 | 30 | 26.7 | 100 | 2.1 |
| Baseline (intervalCap=100) | 20.13.1 | 16,200 | 30 | 7 | 100 | 5.4 |
| Baseline (intervalCap=100) | 20.13.1 | 8,100 | 30 | 26.7 | 100 | 2.7 |
| Baseline (intervalCap=100) | 18.16.1 | 4,200 | 30 | 26.7 | 100 | 1.4 |
| Baseline (intervalCap=5, limit=1000, esCompress=true) | 20.13.1 | 26,000 | 30 | 7 | 1,000 | 0.87 |
| Baseline (intervalCap=5, limit=1000, esCompress=true) | 20.13.1 | 22,000 | 30 | 26.7 | 1,000 | 0.7 |
| Baseline (intervalCap=5) | 18.16.1 | 3,200 | 30 | 26.7 | 100 | 1.06 |

<details>
<summary>Details - Command Lines - Log Outputs</summary>

### This PR with `--concurrency=10` for Writes - `node 20` - `1k Limit` - `esCompress` - `Short RTT`

- Node: 20.13.1
- Records Written: 66,000
- Duration: 30 seconds
- Average RTT (established socket) to Source and Target: 7 ms
- Records per Batch: 100
- Batches / Second = 66,000 / 1,000 / 30 = 2.2

#### Command Line

```sh
npx elasticdump --input=https://source_es/some_index --output=https://target_es/some_index --type=data --input-params="{\"preference\":\"_shards:0\"}" --concurrency=10 --limit=1000 --esCompress=true
```

```log
Fri, 14 Feb 2025 21:08:08 GMT | starting dump
Fri, 14 Feb 2025 21:08:09 GMT | got 1000 objects from source elasticsearch (offset: 0)
Fri, 14 Feb 2025 21:08:09 GMT | got 1000 objects from source elasticsearch (offset: 1000)
Fri, 14 Feb 2025 21:08:09 GMT | sent 1000 objects, 0 offset, to destination elasticsearch, wrote 1000
```

#### End of 30 Seconds of Dump

```log
Fri, 14 Feb 2025 21:08:37 GMT | got 1000 objects from source elasticsearch (offset: 69000)
Fri, 14 Feb 2025 21:08:37 GMT | got 1000 objects from source elasticsearch (offset: 70000)
Fri, 14 Feb 2025 21:08:38 GMT | sent 1000 objects, 65000 offset, to destination elasticsearch, wrote 1000
Fri, 14 Feb 2025 21:08:38 GMT | sent 1000 objects, 66000 offset, to destination elasticsearch, wrote 1000
Fri, 14 Feb 2025 21:08:38 GMT | sent 1000 objects, 67000 offset, to destination elasticsearch, wrote 1000
Fri, 14 Feb 2025 21:08:38 GMT | sent 1000 objects, 68000 offset, to destination elasticsearch, wrote 1000
Fri, 14 Feb 2025 21:08:38 GMT | sent 1000 objects, 69000 offset, to destination elasticsearch, wrote 1000
```

### This PR with `--concurrency=10` for Writes - `node 20` - `1k Limit` - `esCompress`

- Node: 20.13.1
- Records Written: 56,000
- Duration: 30 seconds
- Average RTT (established socket) to Source and Target: 26.7 ms
- Records per Batch: 100
- Batches / Second = 56,000 / 1,000 / 30 = 1.87

#### Command Line

```sh
npx elasticdump --input=https://source_es/some_index --output=https://target_es/some_index --type=data --input-params="{\"preference\":\"_shards:0\"}" --concurrency=10 --limit=1000 --esCompress=true
```

```log
Fri, 14 Feb 2025 20:54:44 GMT | starting dump
Fri, 14 Feb 2025 20:54:45 GMT | got 1000 objects from source elasticsearch (offset: 0)
Fri, 14 Feb 2025 20:54:45 GMT | got 1000 objects from source elasticsearch (offset: 1000)
Fri, 14 Feb 2025 20:54:45 GMT | sent 1000 objects, 0 offset, to destination elasticsearch, wrote 1000
Fri, 14 Feb 2025 20:54:46 GMT | sent 1000 objects, 1000 offset, to destination elasticsearch, wrote 1000
Fri, 14 Feb 2025 20:54:46 GMT | got 1000 objects from source elasticsearch (offset: 2000)
Fri, 14 Feb 2025 20:54:47 GMT | got 1000 objects from source elasticsearch (offset: 3000)
Fri, 14 Feb 2025 20:54:47 GMT | sent 1000 objects, 2000 offset, to destination elasticsearch, wrote 1000
```

#### End of 30 Seconds of Dump

```log
Fri, 14 Feb 2025 20:55:14 GMT | got 1000 objects from source elasticsearch (offset: 55000)
Fri, 14 Feb 2025 20:55:14 GMT | sent 1000 objects, 54000 offset, to destination elasticsearch, wrote 1000
Fri, 14 Feb 2025 20:55:14 GMT | got 1000 objects from source elasticsearch (offset: 56000)
Fri, 14 Feb 2025 20:55:15 GMT | got 1000 objects from source elasticsearch (offset: 57000)
Fri, 14 Feb 2025 20:55:15 GMT | sent 1000 objects, 55000 offset, to destination elasticsearch, wrote 1000
```

### This PR with `--concurrency=10` for Writes - `node 20` - `short RTT`

- Node: 20.13.1
- Records Written: 61,400
- Duration: 30 seconds
- Average RTT (established socket) to Source and Target: 7 ms
- Records per Batch: 100
- Batches / Second = 61,400 / 100 / 30 = 20.5

#### Command Line

```sh
npx elasticdump --input=https://source_es/some_index --output=https://target_es/some_index --type=data --input-params="{\"preference\":\"_shards:0\"}" --concurrency=10
```

```log
Fri, 14 Feb 2025 19:01:21 GMT | starting dump
Fri, 14 Feb 2025 19:01:21 GMT | got 100 objects from source elasticsearch (offset: 0)
Fri, 14 Feb 2025 19:01:21 GMT | got 100 objects from source elasticsearch (offset: 100)
Fri, 14 Feb 2025 19:01:21 GMT | got 100 objects from source elasticsearch (offset: 200)
...
Fri, 14 Feb 2025 19:01:21 GMT | got 100 objects from source elasticsearch (offset: 1800)
Fri, 14 Feb 2025 19:01:21 GMT | sent 100 objects, 800 offset, to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 19:01:21 GMT | got 100 objects from source elasticsearch (offset: 1900)
Fri, 14 Feb 2025 19:01:22 GMT | sent 100 objects, 900 offset, to destination elasticsearch, wrote 100
```

#### End of 30 Seconds of Dump

```log
Fri, 14 Feb 2025 19:01:50 GMT | got 100 objects from source elasticsearch (offset: 61900)
Fri, 14 Feb 2025 19:01:50 GMT | sent 100 objects, 60900 offset, to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 19:01:50 GMT | got 100 objects from source elasticsearch (offset: 62000)
Fri, 14 Feb 2025 19:01:50 GMT | sent 100 objects, 61100 offset, to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 19:01:50 GMT | got 100 objects from source elasticsearch (offset: 62100)
Fri, 14 Feb 2025 19:01:50 GMT | sent 100 objects, 61000 offset, to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 19:01:50 GMT | got 100 objects from source elasticsearch (offset: 62200)
Fri, 14 Feb 2025 19:01:50 GMT | sent 100 objects, 61200 offset, to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 19:01:50 GMT | got 100 objects from source elasticsearch (offset: 62300)
Fri, 14 Feb 2025 19:01:51 GMT | sent 100 objects, 61300 offset, to destination elasticsearch, wrote 100
```

### This PR with `--concurrency=10` for Writes - `node 20`

- Node: 20.13.1
- Records Written: 30,500
- Duration: 30 seconds
- Average RTT (established socket) to Source and Target: 26.7 ms
- Records per Batch: 100
- Batches / Second = 30,500 / 100 / 30 = 10.2

#### Command Line

```sh
npx elasticdump --input=https://source_es/some_index --output=https://target_es/some_index --type=data --input-params="{\"preference\":\"_shards:0\"}" --concurrency=10
```

```log
Fri, 14 Feb 2025 18:29:43 GMT | starting dump
Fri, 14 Feb 2025 18:29:43 GMT | got 100 objects from source elasticsearch (offset: 0)
Fri, 14 Feb 2025 18:29:44 GMT | got 100 objects from source elasticsearch (offset: 100)
Fri, 14 Feb 2025 18:29:44 GMT | got 100 objects from source elasticsearch (offset: 200)
Fri, 14 Feb 2025 18:29:44 GMT | got 100 objects from source elasticsearch (offset: 300)
Fri, 14 Feb 2025 18:29:44 GMT | got 100 objects from source elasticsearch (offset: 400)
Fri, 14 Feb 2025 18:29:44 GMT | got 100 objects from source elasticsearch (offset: 500)
Fri, 14 Feb 2025 18:29:44 GMT | got 100 objects from source elasticsearch (offset: 600)
Fri, 14 Feb 2025 18:29:44 GMT | got 100 objects from source elasticsearch (offset: 700)
Fri, 14 Feb 2025 18:29:44 GMT | sent 100 objects, 500 offset, to destination elasticsearch, wrote 100
```

#### End of 30 Seconds of Dump

```log
Fri, 14 Feb 2025 18:30:13 GMT | sent 100 objects, 30200 offset, to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 18:30:13 GMT | got 100 objects from source elasticsearch (offset: 30500)
Fri, 14 Feb 2025 18:30:13 GMT | got 100 objects from source elasticsearch (offset: 30600)
Fri, 14 Feb 2025 18:30:13 GMT | sent 100 objects, 30300 offset, to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 18:30:14 GMT | got 100 objects from source elasticsearch (offset: 30700)
Fri, 14 Feb 2025 18:30:14 GMT | sent 100 objects, 30400 offset, to destination elasticsearch, wrote 100
```

### This PR with `--concurrency=10` for Writes - `node 18`

- Node: 18.16.1
- Records Written: 13,000
- Duration: 30 seconds
- Average RTT (established socket) to Source and Target: 26.7 ms
- Records per Batch: 100
- Batches / Second = 13,000 / 100 / 30 = 4.3

#### Command Line

```sh
npx elasticdump --input=https://source_es/some_index --output=https://target_es/some_index --type=data --input-params="{\"preference\":\"_shards:0\"}" --concurrency=10
```

```log
Fri, 14 Feb 2025 18:08:06 GMT | starting dump
Fri, 14 Feb 2025 18:08:06 GMT | got 100 objects from source elasticsearch (offset: 0)
Fri, 14 Feb 2025 18:08:07 GMT | got 100 objects from source elasticsearch (offset: 100)
Fri, 14 Feb 2025 18:08:07 GMT | got 100 objects from source elasticsearch (offset: 200)
```

#### End of 30 Seconds of Dump

```log
Fri, 14 Feb 2025 18:08:36 GMT | sent 100 objects, 12700 offset, to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 18:08:36 GMT | got 100 objects from source elasticsearch (offset: 13000)
Fri, 14 Feb 2025 18:08:36 GMT | sent 100 objects, 12800 offset, to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 18:08:37 GMT | sent 100 objects, 12900 offset, to destination elasticsearch, wrote 100
```

### This PR with Default `--concurrency=1` for Writes - `node 20` - `Short RTT`

- Node: 20.13.1
- Records Written: 19,200
- Duration: 30 seconds
- Average RTT (established socket) to Source and Target: 7 ms
- Records per Batch: 100
- Batches / Second = 19,200 / 100 / 30 = 6.4

#### Command Line

```sh
npx elasticdump --input=https://source_es/some_index --output=https://target_es/some_index --type=data --input-params="{\"preference\":\"_shards:0\"}"
```

#### Start of Dump

```log
Fri, 14 Feb 2025 19:05:11 GMT | starting dump
Fri, 14 Feb 2025 19:05:11 GMT | got 100 objects from source elasticsearch (offset: 0)
Fri, 14 Feb 2025 19:05:11 GMT | got 100 objects from source elasticsearch (offset: 100)
Fri, 14 Feb 2025 19:05:11 GMT | sent 100 objects, 0 offset, to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 19:05:11 GMT | got 100 objects from source elasticsearch (offset: 200)
Fri, 14 Feb 2025 19:05:11 GMT | sent 100 objects, 100 offset, to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 19:05:11 GMT | got 100 objects from source elasticsearch (offset: 300)
Fri, 14 Feb 2025 19:05:11 GMT | sent 100 objects, 200 offset, to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 19:05:11 GMT | got 100 objects from source elasticsearch (offset: 400)
Fri, 14 Feb 2025 19:05:12 GMT | sent 100 objects, 300 offset, to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 19:05:12 GMT | got 100 objects from source elasticsearch (offset: 500)
```

#### End of 30 Seconds of Dump

```log
Fri, 14 Feb 2025 19:05:40 GMT | sent 100 objects, 18800 offset, to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 19:05:40 GMT | got 100 objects from source elasticsearch (offset: 19000)
Fri, 14 Feb 2025 19:05:40 GMT | sent 100 objects, 18900 offset, to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 19:05:40 GMT | got 100 objects from source elasticsearch (offset: 19100)
Fri, 14 Feb 2025 19:05:40 GMT | sent 100 objects, 19000 offset, to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 19:05:40 GMT | got 100 objects from source elasticsearch (offset: 19200)
Fri, 14 Feb 2025 19:05:41 GMT | sent 100 objects, 19100 offset, to destination elasticsearch, wrote 100
```

### This PR with Default `--concurrency=1` for Writes - `node 20`

- Node: 20.13.1
- Records Written: 10,500
- Duration: 30 seconds
- Average RTT (established socket) to Source and Target: 26.7 ms
- Records per Batch: 100
- Batches / Second = 10,500 / 100 / 30 = 3.5

#### Command Line

```sh
npx elasticdump --input=https://source_es/some_index --output=https://target_es/some_index --type=data --input-params="{\"preference\":\"_shards:0\"}"
```

#### Start of Dump

```log
Fri, 14 Feb 2025 18:32:44 GMT | starting dump
Fri, 14 Feb 2025 18:32:44 GMT | got 100 objects from source elasticsearch (offset: 0)
Fri, 14 Feb 2025 18:32:44 GMT | got 100 objects from source elasticsearch (offset: 100)
Fri, 14 Feb 2025 18:32:45 GMT | sent 100 objects, 0 offset, to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 18:32:45 GMT | got 100 objects from source elasticsearch (offset: 200)
```

#### End of 30 Seconds of Dump

```log
Fri, 14 Feb 2025 18:33:14 GMT | sent 100 objects, 10200 offset, to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 18:33:14 GMT | got 100 objects from source elasticsearch (offset: 10400)
Fri, 14 Feb 2025 18:33:14 GMT | sent 100 objects, 10300 offset, to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 18:33:14 GMT | got 100 objects from source elasticsearch (offset: 10500)
Fri, 14 Feb 2025 18:33:15 GMT | sent 100 objects, 10400 offset, to destination elasticsearch, wrote 100
```

### This PR with Default `--concurrency=1` for Writes - `node 18`

- Node: 18.16.1
- Records Written: 6,300
- Duration: 30 seconds
- Average RTT (established socket) to Source and Target: 26.7 ms
- Records per Batch: 100
- Batches / Second = 6,300 / 100 / 30 = 2.1

#### Command Line

```sh
npx elasticdump --input=https://source_es/some_index --output=https://target_es/some_index --type=data --input-params="{\"preference\":\"_shards:0\"}"
```

#### Start of Dump

```log
Fri, 14 Feb 2025 17:56:24 GMT | starting dump
Fri, 14 Feb 2025 17:56:25 GMT | got 100 objects from source elasticsearch (offset: 0)
Fri, 14 Feb 2025 17:56:25 GMT | got 100 objects from source elasticsearch (offset: 100)
Fri, 14 Feb 2025 17:56:26 GMT | sent 100 objects, 0 offset, to destination elasticsearch, wrote 100
```

#### End of 30 Seconds of Dump

```log
Fri, 14 Feb 2025 17:56:54 GMT | got 100 objects from source elasticsearch (offset: 6200)
Fri, 14 Feb 2025 17:56:54 GMT | sent 100 objects, 6100 offset, to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 17:56:54 GMT | got 100 objects from source elasticsearch (offset: 6300)
Fri, 14 Feb 2025 17:56:55 GMT | sent 100 objects, 6200 offset, to destination elasticsearch, wrote 100
```

### Baseline of `6.117.0` with `--intervalCap=100` - `node 20` - `Short RTT`

- Node: 20.13.1
- Records Written: 16,200
- Duration: 30 seconds
- Average RTT (established socket) to Source and Target: 7 ms
- Records per Batch: 100
- Batches / Second = 16,200 / 100 / 30 = 5.4

#### Command Line

```sh
elasticdump --input=https://source_es/some_index --output=https://target_es/some_index --type=data --input-params="{\"preference\":\"_shards:0\"}" --intervalCap=100 
```

#### Start of Dump

```log
Fri, 14 Feb 2025 19:33:32 GMT | starting dump
Fri, 14 Feb 2025 19:33:32 GMT | got 100 objects from source elasticsearch (offset: 0)
Fri, 14 Feb 2025 19:33:32 GMT | sent 100 objects to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 19:33:32 GMT | got 100 objects from source elasticsearch (offset: 100)
Fri, 14 Feb 2025 19:33:32 GMT | sent 100 objects to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 19:33:32 GMT | got 100 objects from source elasticsearch (offset: 200)
Fri, 14 Feb 2025 19:33:32 GMT | sent 100 objects to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 19:33:33 GMT | got 100 objects from source elasticsearch (offset: 300)
Fri, 14 Feb 2025 19:33:33 GMT | sent 100 objects to destination elasticsearch, wrote 100
```

#### End of 30 Seconds of Dump

```log
Fri, 14 Feb 2025 19:34:01 GMT | sent 100 objects to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 19:34:01 GMT | got 100 objects from source elasticsearch (offset: 15900)
Fri, 14 Feb 2025 19:34:01 GMT | sent 100 objects to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 19:34:01 GMT | got 100 objects from source elasticsearch (offset: 16000)
Fri, 14 Feb 2025 19:34:01 GMT | sent 100 objects to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 19:34:01 GMT | got 100 objects from source elasticsearch (offset: 16100)
Fri, 14 Feb 2025 19:34:02 GMT | sent 100 objects to destination elasticsearch, wrote 100
```

### Baseline of `6.117.0` with `--intervalCap=100` - `node 20`

- Node: 20.13.1
- Records Written: 8,100
- Duration: 30 seconds
- Average RTT (established socket) to Source and Target: 26.7 ms
- Records per Batch: 100
- Batches / Second = 8,100 / 100 / 30 = 2.7

#### Command Line

```sh
elasticdump --input=https://source_es/some_index --output=https://target_es/some_index --type=data --input-params="{\"preference\":\"_shards:0\"}" --intervalCap=100 
```

#### Start of Dump

```log
Fri, 14 Feb 2025 18:25:00 GMT | starting dump
Fri, 14 Feb 2025 18:25:00 GMT | got 100 objects from source elasticsearch (offset: 0)
Fri, 14 Feb 2025 18:25:01 GMT | sent 100 objects to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 18:25:01 GMT | got 100 objects from source elasticsearch (offset: 100)
```

#### End of 30 Seconds of Dump

```log
Fri, 14 Feb 2025 18:25:30 GMT | got 100 objects from source elasticsearch (offset: 7900)
Fri, 14 Feb 2025 18:25:30 GMT | sent 100 objects to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 18:25:30 GMT | got 100 objects from source elasticsearch (offset: 8000)
Fri, 14 Feb 2025 18:25:31 GMT | sent 100 objects to destination elasticsearch, wrote 100
```

### Baseline of `6.117.0` with `--intervalCap=100` - `node 18`

- Node: 18.16.1
- Records Written: 4,200
- Duration: 30 seconds
- Average RTT (established socket) to Source and Target: 26.7 ms
- Records per Batch: 100
- Batches / Second = 4,200 / 100 / 30 = 1.4

#### Command Line

```sh
elasticdump --input=https://source_es/some_index --output=https://target_es/some_index --type=data --input-params="{\"preference\":\"_shards:0\"}" --intervalCap=100 
```

#### Start of Dump

```log
Fri, 14 Feb 2025 18:15:29 GMT | starting dump
Fri, 14 Feb 2025 18:15:29 GMT | got 100 objects from source elasticsearch (offset: 0)
Fri, 14 Feb 2025 18:15:30 GMT | sent 100 objects to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 18:15:30 GMT | got 100 objects from source elasticsearch (offset: 100)
```

#### End of 30 Seconds of Dump

```log
Fri, 14 Feb 2025 18:15:58 GMT | sent 100 objects to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 18:15:58 GMT | got 100 objects from source elasticsearch (offset: 4100)
Fri, 14 Feb 2025 18:15:59 GMT | sent 100 objects to destination elasticsearch, wrote 100
```

### Baseline of `6.117.0` with Default `--intervalCap=5` - `node 20` - `esCompress` - `Short RTT`

- Node: 20.13.1
- Records Written: 26,000
- Duration: 30 seconds
- Average RTT (established socket) to Source and Target: 7 ms
- Records per Batch: 1,000
- Batches / Second = 26,000 / 1,000 / 30 = 0.87

#### Command Line

```sh
elasticdump --input=https://source_es/some_index --output=https://target_es/some_index --type=data --input-params="{\"preference\":\"_shards:0\"}" --limit 1000 --esCompress=1000
```

#### Start of Dump

```log
Fri, 14 Feb 2025 21:13:28 GMT | starting dump
Fri, 14 Feb 2025 21:13:28 GMT | got 1000 objects from source elasticsearch (offset: 0)
Fri, 14 Feb 2025 21:13:29 GMT | sent 1000 objects to destination elasticsearch, wrote 1000
Fri, 14 Feb 2025 21:13:29 GMT | got 1000 objects from source elasticsearch (offset: 1000)
```

#### End of 30 Seconds of Dump

```log
Fri, 14 Feb 2025 21:13:56 GMT | got 1000 objects from source elasticsearch (offset: 25000)
Fri, 14 Feb 2025 21:13:57 GMT | sent 1000 objects to destination elasticsearch, wrote 1000
Fri, 14 Feb 2025 21:13:58 GMT | got 1000 objects from source elasticsearch (offset: 26000)
Fri, 14 Feb 2025 21:13:58 GMT | sent 1000 objects to destination elasticsearch, wrote 1000
```

### Baseline of `6.117.0` with Default `--intervalCap=5` - `node 20` - `esCompress`

- Node: 18.16.1
- Records Written: 22,000
- Duration: 30 seconds
- Average RTT (established socket) to Source and Target: 26.7 ms
- Records per Batch: 1,000
- Batches / Second = 22,000 / 1,000 / 30 = 0.73

#### Command Line

```sh
elasticdump --input=https://source_es/some_index --output=https://target_es/some_index --type=data --input-params="{\"preference\":\"_shards:0\"}" --limit 1000 --esCompress=1000
```

#### Start of Dump

```log
Fri, 14 Feb 2025 21:04:21 GMT | starting dump
Fri, 14 Feb 2025 21:04:22 GMT | got 1000 objects from source elasticsearch (offset: 0)
Fri, 14 Feb 2025 21:04:23 GMT | sent 1000 objects to destination elasticsearch, wrote 1000
```

#### End of 30 Seconds of Dump

```log
Fri, 14 Feb 2025 21:04:50 GMT | got 1000 objects from source elasticsearch (offset: 21000)
Fri, 14 Feb 2025 21:04:51 GMT | sent 1000 objects to destination elasticsearch, wrote 1000
```

### Baseline of `6.117.0` with Default `--intervalCap=5` - `node 18`

- Node: 18.16.1
- Records Written: 3,200
- Duration: 30 seconds
- Average RTT (established socket) to Source and Target: 26.7 ms
- Records per Batch: 100
- Batches / Second = 3,200 / 100 / 30 = 1.06

#### Command Line

```sh
elasticdump --input=https://source_es/some_index --output=https://target_es/some_index --type=data --input-params="{\"preference\":\"_shards:0\"}"
```

#### Start of Dump

```log
Fri, 14 Feb 2025 18:12:18 GMT | starting dump
Fri, 14 Feb 2025 18:12:18 GMT | got 100 objects from source elasticsearch (offset: 0)
Fri, 14 Feb 2025 18:12:19 GMT | sent 100 objects to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 18:12:19 GMT | got 100 objects from source elasticsearch (offset: 100)
```

#### End of 30 Seconds of Dump

```log
Fri, 14 Feb 2025 18:12:47 GMT | sent 100 objects to destination elasticsearch, wrote 100
Fri, 14 Feb 2025 18:12:48 GMT | got 100 objects from source elasticsearch (offset: 3100)
Fri, 14 Feb 2025 18:12:49 GMT | sent 100 objects to destination elasticsearch, wrote 100
```

</details>

## Confirmation of Promises Leak Fix

<img width="1169" alt="image" src="https://github.com/user-attachments/assets/5b6797e6-1364-4e9a-9573-e5e7c28b88f3" />

ferronrsmith commented:
This is awesome 😄

ferronrsmith commented:
Everything is merged, you can rebase here and continue.

ferronrsmith commented:
Based on your performance testing. I think I will do a major release v7 and bump the version requirement to node v20. 

Let me know what you think ?

huntharo commented:
> Everything is merged, you can rebase here and continue.
> 
> 

I like that. I'll rebase and get this ready in about an hour.

huntharo commented:
@ferronrsmith I have rebased, added a test and fix to confirm that the destination close happens after all data writes (failing before fix when `concurrency > 1`), taken a final look through the code, and I think this is ready to go whenever you are.

ferronrsmith commented:
@huntharo merged and released https://github.com/elasticsearch-dump/elasticsearch-dump/releases/tag/v6.119.0
